### PR TITLE
support for data migration of incremental snaps on xen

### PR DESCRIPTION
**Pull request by Pearl1594**
## Description
PR attempts to fix and enhance the following:
- On Xen, there are instances when the child snapshot is basically only a database entry with the file path pointing to the parent snapshot - handle this scenario such that a new entry is created for such incremental snapshots and delete the snapshot from source datastore during data migration
- Wrt scaling down of SSVM, revised the logic to prevent quick scaling down of the SSVMs

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?
- Conducted tests on xen of files 
Tests on xen for snapshots with child snaps having install path same as that as parent snapshot
```
Before Migration:
-----------------------

MariaDB [cloud]> select * from snapshot_store_ref where store_role = 'Image' and volume_id =11 and state <> 'Destroyed' order by parent_snapshot_id\G
*************************** 1. row ***************************
                id: 497
          store_id: 1
       snapshot_id: 63
           created: 2020-10-10 07:00:08
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 52531712
parent_snapshot_id: 0
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-10 13:25:24
         volume_id: 11
*************************** 2. row ***************************
                id: 375
          store_id: 1
       snapshot_id: 67
           created: 2020-10-10 07:00:09
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 63
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-08 06:54:57
         volume_id: 11
*************************** 3. row ***************************
                id: 381
          store_id: 1
       snapshot_id: 70
           created: 2020-10-10 07:00:11
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 67
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-08 06:55:14
         volume_id: 11
*************************** 4. row ***************************
                id: 472
          store_id: 1
       snapshot_id: 74
           created: 2020-10-10 07:32:57
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 70
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-10 07:32:57
         volume_id: 11

After migration:
---------------

MariaDB [cloud]> select * from snapshot_store_ref where store_role = 'Image' and volume_id =11 and state <> 'Destroyed' order by parent_snapshot_id\G
*************************** 1. row ***************************
                id: 510
          store_id: 5
       snapshot_id: 63
           created: 2020-10-10 07:00:08
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 52531712
parent_snapshot_id: 0
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-12 04:11:10
         volume_id: 11
*************************** 2. row ***************************
                id: 375
          store_id: 5
       snapshot_id: 67
           created: 2020-10-10 07:00:09
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 63
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-08 06:54:57
         volume_id: 11
*************************** 3. row ***************************
                id: 381
          store_id: 5
       snapshot_id: 70
           created: 2020-10-10 07:00:11
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 67
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-08 06:55:14
         volume_id: 11
*************************** 4. row ***************************
                id: 472
          store_id: 5
       snapshot_id: 74
           created: 2020-10-10 07:32:57
      last_updated: NULL
            job_id: NULL
        store_role: Image
              size: 52428800
     physical_size: 0
parent_snapshot_id: 70
      install_path: snapshots/2/11/d05a2c2f-6885-4932-b5a0-d11f82129a48.vhd
             state: Ready
      update_count: 2
           ref_cnt: 0
           updated: 2020-10-10 07:32:57
         volume_id: 11


```

Tested on xen, vmware and KVM
Tests for SSVM scale down:
Updated following global setting values:
secstorage.capacity.standby: 4
secstorage.session.max: 10

1. Added 7 entries to the cmd_exec_log table, to simulate scaling up operation - Successfully scaled - the new SSVM spawned continues to exist until the number of cmds drops below threshold
2. Deleted entries from cmd_exec_log table, such that only 4 entries were present in the table - successfully triggered scale down - as the total number of cmds in the table has dropped below half the defined max limit (10/2 = 5)
3. Simulated another scale up operation by adding entries to the cmd_exec log table, added entries to the second (new) SSVM however the total number of commands in the table is below threshold - does NOT scale down SSVM as a command is running on it - as expected

### CODE REVIEW

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test matrix

**Comment by rhtyd** [comment]
cc @vladimirpetrov @PaulAngus

**Comment by DaanHoogland** [comment]
@blueorangutan centos7 kvm-centos7 keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan centos7 xenserver-71 keepEnv

**Comment by DaanHoogland** [comment]
@blueorangutan centos7 xenserver-71 keepEnv

**Comment by rhtyd** [comment]
Tests LGTM

**Comment by DaanHoogland** [comment]
@blueorangutan centos7 xenserver-71 keepEnv

**Comment by harikrishna-patnala** [comment]
Code changes LGTM

**Review by harikrishna-patnala — APPROVED** [review]

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/SecondaryStorageServiceImpl.java [hunk 1]
```diff
@@ -198,6 +204,39 @@ protected Void migrateDataCallBack(AsyncCallbackDispatcher<SecondaryStorageServi
...
+            VolumeDataStoreVO srcVolume = volumeDataStoreDao.findByStoreVolume(srcData.getDataStore().getId(), srcData.getId());
+            VolumeDataStoreVO destVolume = volumeDataStoreDao.findByStoreVolume(destData.getDataStore().getId(), destData.getId());
+            if (srcVolume != null && destVolume != null) {
+                destVolume.setPhysicalSize(srcVolume.getPhysicalSize());
+                destVolume.setCreated(srcVolume.getCreated());
+                volumeDataStoreDao.update(destVolume.getId(), destVolume);
+            }
+        }
+
+        if (destData instanceof TemplateInfo) {
```
Comment (sureshanaparti) [inline engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/SecondaryStorageServiceImpl.java:231]: @Pearl1594 better use if-else / switch-case. The control will come here even when `destData` is of type SnapshotInfo. If this method is being called for any other destData instance, better add a debug log for unsupported instance.

File: engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java [hunk 2]
```diff
@@ -391,12 +391,16 @@ public void copyAsync(DataObject srcdata, DataObject destData, AsyncCompletionCa
     private Answer sendToLeastBusyEndpoint(List<EndPoint> eps, CopyCommand cmd) {
         Answer answer = null;
         EndPoint endPoint = null;
-        Long epId = ssvmWithLeastMigrateJobs();
+        List<Long> epIds = ssvmWithLeastMigrateJobs();
+        Long epId = null;
+        if (!epIds.isEmpty()) {
+            epId = epIds.get(0);
+        }
         if (epId == null) {
```
Comment (sureshanaparti) [inline engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java:399]: @Pearl1594 seems `epId` is no longer required for the check here. Can check with empty `epIds` ?

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 3]
```diff
@@ -1309,6 +1309,8 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
                 File srcDir = null;
                 if (srcFile.isFile() || srcFile.getName().contains(".")) {
                     srcDir = new File(srcFile.getParent());
+                } else if (!srcFile.isFile() && !srcFile.isDirectory()) {
```
Comment (sureshanaparti) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1312]: `!srcFile.isFile()` is always true here

File: engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/SecondaryStorageServiceImpl.java (hunk 1 above)
Comment (DaanHoogland) [reply engine/storage/image/src/main/java/org/apache/cloudstack/storage/image/SecondaryStorageServiceImpl.java:231]: agree, it might also be usefull to extract the processing bits in a (template) method to reduce complexity. (no blocker)

File: engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java (hunk 2 above)
Comment (DaanHoogland) [reply engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java:399]: @sureshanaparti is this a remark to style? if epIds is empty epId is null so the check is needed in some way.

**Review by DaanHoogland — APPROVED** [review]
a few locations where complexity is introduced. I would like to see it reduced but overall fine with the code

File: engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java (hunk 2 above)
Comment (sureshanaparti) [reply engine/storage/src/main/java/org/apache/cloudstack/storage/image/BaseImageStoreDriverImpl.java:399]: @DaanHoogland the last method _ssvmWithLeastMigrateJobs()_ returns null if it fails to pick ssvm and the null check seems to be valid. The same updated method now returns an empty list if fails. So list with empty check is fine, 'epId' may not be required.

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 4]
```diff
@@ -1300,16 +1300,16 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
         try {
             File srcFile = new File(getDir(srcStore.getUrl(), _nfsVersion), srcData.getPath());
             File destFile = new File(getDir(destStore.getUrl(), _nfsVersion), destData.getPath());
-            ImageFormat format = getTemplateFormat(srcFile.getName());
 
             if (srcFile == null) {
-                return new CopyCmdAnswer("Can't find src file:" + srcFile);
+                return new CopyCmdAnswer("Can't find src file");
```
Comment (DaanHoogland) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1305]: maybe rephrase to "no src file given/to work with" or something?

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 5]
```diff
@@ -1300,16 +1300,16 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
...
             }
+            ImageFormat format = getTemplateFormat(srcFile.getName());
             if (srcData instanceof TemplateObjectTO || srcData instanceof VolumeObjectTO) {
                 File srcDir = null;
                 if (srcFile.isFile() || srcFile.getName().contains(".")) {
                     srcDir = new File(srcFile.getParent());
-                } else if (!srcFile.isFile() && !srcFile.isDirectory()) {
+                } else if (!srcFile.isDirectory()) {
```
Comment (DaanHoogland) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1312]: is this supposed to be **!**isDir()? or isDir() plain?

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan package

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 6]
```diff
@@ -1300,15 +1300,17 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
         try {
             File srcFile = new File(getDir(srcStore.getUrl(), _nfsVersion), srcData.getPath());
             File destFile = new File(getDir(destStore.getUrl(), _nfsVersion), destData.getPath());
-            ImageFormat format = getTemplateFormat(srcFile.getName());
 
             if (srcFile == null) {
-                return new CopyCmdAnswer("Can't find src file:" + srcFile);
+                return new CopyCmdAnswer("Can't find source file to initiate file transfer");
```
Comment (sureshanaparti) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1305]: adding path and datastore id to the log might help in debugging

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 7]
```diff
@@ -1351,7 +1353,7 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
                 if (srcFile.isFile()) {
                     newVol.setPath(destData.getPath() + File.separator + srcFile.getName());
                 } else {
-                    newVol.setPath(destData.getPath());
+                    newVol.setPath(srcData.getPath());
```
Comment (sureshanaparti) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1356]: check if this causes any regression

Comment (Pearl1594) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1356]: Have verified this

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java (hunk 6 above)
Comment (Pearl1594) [reply services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java]: done

**Comment by Pearl1594** [comment]
@blueorangutan test

File: services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java [hunk 8]
```diff
@@ -1300,15 +1300,17 @@ protected Answer copyFromNfsToNfs(CopyCommand cmd) {
         try {
             File srcFile = new File(getDir(srcStore.getUrl(), _nfsVersion), srcData.getPath());
             File destFile = new File(getDir(destStore.getUrl(), _nfsVersion), destData.getPath());
-            ImageFormat format = getTemplateFormat(srcFile.getName());
 
             if (srcFile == null) {
-                return new CopyCmdAnswer("Can't find src file:" + srcFile);
+                return new CopyCmdAnswer("Can't find source file at path: "+ srcData.getPath() +"on datastore: "+ srcDataStore.getUuid() +" to initiate file transfer");
```
Comment (sureshanaparti) [inline services/secondary-storage/server/src/main/java/org/apache/cloudstack/storage/resource/NfsSecondaryStorageResource.java:1305]: make sure spaces are properly aligned on the log code, rest changes LGTM

**Review by sureshanaparti — APPROVED** [review]
